[Form][FrameworkBundle][Bridge] Add a DateInterval form type

MisatoTremor opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13389 |
| License | MIT |
| Doc PR | symfony/symfony-docs#4817 |

Replaces #13390

MisatoTremor committed:
[Form] Add a DateInterval form type
Also add dateinterval widget to twig templates.

OskarStark commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php:
```diff
@@ -0,0 +1,171 @@
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
```
is the header copied from other files?

i think you should be named here @MisatoTremor

MisatoTremor replied:
Yes, this is the standard copyright header.
I've put my name at the class author PHPDoc as usual. ;)

OskarStark replied:
ok, i didn't know.

thank you!

jewome62 commented:
:cool:  !!!

When that will be integrated into the 2.8?

stof commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php:
```diff
@@ -0,0 +1,127 @@
+            throw new TransformationFailedException('Expected a string.');
+        }
+        try {
+            if ($this->isISO8601($value)) {
```
IMO, this if/else should be outside the try/catch, avoiding the need to catch TransformationFailedException as is (and it should use an early return)

MisatoTremor replied:
Good catch, makes absolute sense to put it outside.
But is it really OK to just return null if the value is something the transformer does not expect?

MisatoTremor replied:
@stof 
Just saw again that this throw > catch > throw pattern is also used in DateTimeToStringTransformer. Most probably I got it originally from there.

So maybe both should be changed (DateTimeToStringTransformer in another PR then) or just be left as is.

What do you think?

stof replied:
Simplifying the DateTimeToStringTransformer in another PR makes sense yeah. but it should indeed be a separate PR

MisatoTremor commented:
@stof Ok, simplified the transformer and also moved the format check out of the try/catch as this also was a throw > catch > throw.
Also fixed the deprecated form type by string name references and squashed commits.

AppVeyor build failure seems because of something else, as it also fails for the same tests in other PRs, etc. and Form component test passes.

sstok commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php:
```diff
@@ -0,0 +1,171 @@
+class DateIntervalToArrayTransformer implements DataTransformerInterface
+{
+    private $fields;
+    private $availableFields = array(
```
This can be made static as it actually never changes.

Same done in https://github.com/MisatoTremor/symfony/blob/add_dateinterval_type_2_8/src/Symfony/Component/Form/Extension/Core/DataTransformer/PercentToLocalizedStringTransformer.php#L29

sstok commented on src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php:
```diff
@@ -0,0 +1,124 @@
+            throw new TransformationFailedException('Expected a string.');
+        }
+        if (!$this->isISO8601($value)) {
+            throw new TransformationFailedException('Non ISO 8601 date strings are not supported yet');
```
What about: `Only date intervals in the ISO 8601 format are supported.`

sstok commented on src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/DateIntervalTestCase.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Symfony\Component\Form\Tests\Extension\Core\DataTransformer;
+
+class DateIntervalTestCase extends \PHPUnit_Framework_TestCase
```
Should be marked abstract to prevent being executed as test-class.

MisatoTremor committed:
Mark DateIntervalTestCase abstract, make available fields in DateIntervalToArrayTransformer static

MisatoTremor commented:
Rebased on master
